lib/trie/version.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package trie
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// Version is the state trie version which dictates how a
+// Merkle root should be constructed. It is defined in
+// https://spec.polkadot.network/#defn-state-version
+type Version uint8
+
+const (
+	// V0 is the state trie version 0 where the values of the keys are
+	// inserted into the trie directly.
+	// TODO set to iota once CI passes
+	V0 Version = 1
+)
+
+func (v Version) String() string {
+	switch v {
+	case V0:
+		return "v0"
+	default:
+		panic(fmt.Sprintf("unknown version %d", v))
+	}
+}
+
+var ErrParseVersion = errors.New("parsing version failed")
+
+// ParseVersion parses a state trie version string.
+func ParseVersion(s string) (version Version, err error) {
+	switch {
+	case strings.EqualFold(s, V0.String()):
+		return V0, nil
+	default:
+		choices := orStrings([]string{V0.String()})
+		return version, fmt.Errorf("%w: %q must be %s",
+			ErrParseVersion, s, choices)
```

Comment:
I believe the usage of `strings.Join` could be enough to show the amount of options you have to `ParseVersion` and more simple to understand:
```suggestion
		options := strings.Join([]string{V0.String()}, ", ")
		return version, fmt.Errorf("%w: %q should be one of %s",
			ErrParseVersion, s, options)
```

Reply:
I midagree (disagree and agree!)

Re-reading this, this is just brainiac thoughts spreaded in code, and simple code is important.
BUT it's also nice to see `"V2" must be V0 or V1` instead of `"V2" should be one of V0, V1`.

Now since it makes you happy, let's go with your suggestion 😉 Change to:

```go
		options := strings.Join([]string{V0.String()}, ", ")
		return version, fmt.Errorf("%w: %q must be one of %s",
			ErrParseVersion, s, options)
```

Reply:
I actually just changed a bit to be `fmt.Errorf("%w: %q must be %s", ErrParseVersion, s, V0)` since there is only v0 for now. I'll update the error to be `fmt.Errorf("%w: %q must be %s or %s", ErrParseVersion, s, V0, V1)` when adding V1, otherwise error messages are just plain weird.